docs/eBpfExtensions.md

```diff
@@ -1,42 +1,207 @@
+2. Authoring Helper Functions.
+3. Invoking eBPF programs from hook(s).
+4. Registering program and attach types.
+
+The following sections describe these steps in detail.
 
-Currently the information is exposed as an opaque blob of
-[MS-RPCE](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rpce/290c38b1-92fe-4229-91e6-4fc376610c15)
-serialized data with the schema defined by
-[ebpf_program_types](https://github.com/microsoft/ebpf-for-windows/blob/master/libs/platform/ebpf_program_types.idl).
-Due to current limitations on Windows, it's not possible to serialize the contract from kernel mode. Proposals are
-underway to switch the serialization to use [Google FlatBuffers](https://google.github.io/flatbuffers/) as that is
-more easily serialized from kernel mode.
+### 2.1 Program Information NPI Provider Registration
+When registering itself to the NMR, the Program Information NPI Provider should have the [`NPI_REGISTRATION_INSTANCE`](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/netioddk/ns-netioddk-_npi_registration_instance) initialized as follows:
+* `NpiId`: This should be set the GUID corresponding to the eBPF program type.
+* `ModuleId`: This should be a unique GUID identifying the provider module.
+* `NpiSpecificCharacteristics`: Pointer to structure of type `ebpf_extension_data_t`.
+  * The `data` field of this structure should point to a structure of type `ebpf_program_data_t`.
 
-## NPI Contract for Attach Type
+#### `ebpf_program_data_t` Struct
+The various fields of this structure should be set as follows:
+* `program_info`: Pointer to `ebpf_program_info_t`.
+* `helper_function_addresses`: Pointer to `ebpf_helper_function_addresses_t`.
 
-Attach type NPI contracts are the mechanism that extensions use to invoke eBPF programs when events occur. The
-eBPF extension registers as an NMR provider and supplies callback functions that are invoked when a NMR client
-attaches or detaches. To invoke an eBPF program, the extension uses the NpiSpecificCharacteristics as well as the
-supplied dispatch table. The first entry in the dispatch table is a pointer to the function that invokes the eBPF
-program and has the signature:
+#### `ebpf_program_info_t` Struct
+The various fields of this structure should be set as follows:
+* `program_type_descriptor`: Field of type `ebpf_program_type_descriptor_t`.
+* `count_of_helpers`: The number of helper function that are implemented by this extension for the given program type.
+* `helper_prototype`: Pointer to array of `ebpf_helper_function_prototype_t`.
 
+#### `ebpf_program_type_descriptor_t` Struct
+The various fields of this structure should be set as follows:
+* `name`: Friendly name of the program type.
+* `context_descriptor`: Pointer of type `ebpf_context_descriptor_t`.
+* `program_type`: GUID for the program type. This should also be the NPI ID.
+* `is_privileged`: This Boolean field should be set to `TRUE` if this is a privileged program type, `FALSE` otherwise.
```

Comment:
Can you define what "privileged program type" means here?

Reply:
In PREVAIL, the only privileged program type is BPF_PROG_TYPE_KPROBE.
I also pinged Elazar to try to get a description.

Reply:
Changed to set to FALSE for now. Will revisit later.